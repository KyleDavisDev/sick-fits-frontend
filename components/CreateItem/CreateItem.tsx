import * as React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import formatMoney from "../../lib/formatMoney";

interface ICreateItem {}

class CreateItem extends React.Component<ICreateItem, {}> {
  public render() {
    return <p>CreateItems</p>;
  }
}

export default CreateItem;
