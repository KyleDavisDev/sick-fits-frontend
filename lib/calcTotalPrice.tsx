export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
