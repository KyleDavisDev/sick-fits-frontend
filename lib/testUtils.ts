import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
