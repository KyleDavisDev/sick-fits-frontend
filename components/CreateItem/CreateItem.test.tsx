import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import CreateItem, { CREATE_ITEM_MUTATION } from "./CreateItem";
import { fakeItem } from "../../lib/testUtils";

// mock the global fetch API
const dogImage = "https://dog.com/dog.jpg";

describe("<CreateItem />", () => {
  it("renders", () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    expect(wrapper).toBeTruthy();
  });

  it("matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    const form = wrapper.find("form");
    expect(form).toMatchSnapshot();
  });

  it("uploads an image when changed", async () => {
    // mock fetch
    // @ts-ignore
    global.fetch = jest.fn().mockResolvedValue({
      json: () => ({
        secure_url: dogImage,
        eager: [{ secure_url: dogImage }]
      })
    });

    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    const input = wrapper.find("input[type='file']");
    input.simulate("change", { target: { files: [dogImage] } });

    // let component update/load
    await wait();
    wrapper.update();

    // grab component
    const component = wrapper.find("CreateItem").instance() as CreateItem;

    // check state
    expect(component.state.image).toEqual(dogImage);
    expect(component.state.largeImage).toEqual(dogImage);
    // @ts-ignore
    expect(global.fetch).toHaveBeenCalled();

    // reset the global fetch
    // @ts-ignore
    global.fetch.mockReset();
  });
});
