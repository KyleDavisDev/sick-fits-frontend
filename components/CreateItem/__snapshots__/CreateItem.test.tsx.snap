// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> matches snapshot 1`] = `
<form
  className="sc-bdVaJa cgcGkz"
  method="post"
  onSubmit={[Function]}
>
  <ErrorMessage
    error={Object {}}
  />
  <h2>
    CreateItem an Item
  </h2>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Upload File
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image..."
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        required={true}
        type="number"
        value={15}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a description..."
        required={true}
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
